api/accessrequest/access_request.go

```diff
@@ -0,0 +1,177 @@
+)
+
+type ListResourcesRequestOption func(*proto.ListResourcesRequest)
+
+// GetResourcesWithFilters is a wrapper for client.GetResourcesWithFilters
+// that maps resource kinds to the resource types expected by ListResources.
+func GetResourcesWithFilters(ctx context.Context, clt client.ListResourcesClient, req proto.ListResourcesRequest, kind string) ([]types.ResourceWithLabels, error) {
+	req.ResourceType = mapResourceKindToListResourcesType(kind)
+	results, err := client.GetResourcesWithFilters(ctx, clt, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resources := make([]types.ResourceWithLabels, 0, len(results))
+	for _, result := range results {
+		leafResource, err := mapListResourcesResultToLeafResource(result, kind)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		resources = append(resources, leafResource)
+	}
+	return resources, nil
+}
+
+// GetResourceDetails gets extra details for a list of resources in a given cluster.
+func GetResourceDetails(ctx context.Context, clusterName string, lister client.ListResourcesClient, ids []types.ResourceID) (map[string]types.ResourceDetails, error) {
+	var resourceIDs []types.ResourceID
+	for _, resourceID := range ids {
+		// We're interested in hostname or friendly name details. These apply to
+		// nodes, app servers, and user groups.
+		switch resourceID.Kind {
+		case types.KindNode, types.KindApp, types.KindUserGroup:
+			resourceIDs = append(resourceIDs, resourceID)
+		}
+	}
+
+	withExtraRoles := func(req *proto.ListResourcesRequest) {
+		req.UseSearchAsRoles = true
+		req.UsePreviewAsRoles = true
+	}
+
+	resources, err := GetResourcesByResourceIDs(ctx, lister, resourceIDs, withExtraRoles)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	result := make(map[string]types.ResourceDetails)
+	for _, resource := range resources {
+		friendlyName := types.FriendlyName(resource)
+
+		// No friendly name was found, so skip to the next resource.
+		if friendlyName == "" {
+			continue
+		}
+
+		id := types.ResourceID{
+			ClusterName: clusterName,
+			Kind:        resource.GetKind(),
+			Name:        resource.GetName(),
+		}
+		result[types.ResourceIDToString(id)] = types.ResourceDetails{
+			FriendlyName: friendlyName,
+		}
+	}
+
+	return result, nil
+}
+
+// GetResourceIDsByCluster will return resource IDs grouped by cluster.
+func GetResourceIDsByCluster(r types.AccessRequest) map[string][]types.ResourceID {
+	resourceIDsByCluster := make(map[string][]types.ResourceID)
+	for _, resourceID := range r.GetRequestedResourceIDs() {
+		resourceIDsByCluster[resourceID.ClusterName] = append(resourceIDsByCluster[resourceID.ClusterName], resourceID)
+	}
+	return resourceIDsByCluster
+}
+
+// GetResourcesByResourceID gets a list of resources by their resource IDs.
+func GetResourcesByResourceIDs(ctx context.Context, lister client.ListResourcesClient, resourceIDs []types.ResourceID, opts ...ListResourcesRequestOption) ([]types.ResourceWithLabels, error) {
```

Comment:
Is this ever called outside this package? If not, let's unexport it.

Reply:
It's called in lib/services.